pints/_mcmc/_slice_stepout.py

```diff
@@ -0,0 +1,430 @@
+
+    [1] Neal, R.M., 2003. Slice sampling. The annals of statistics, 31(3),
+    pp.705-767.
+    """
+
+    def __init__(self, x0, sigma0=None):
+        super(SliceStepoutMCMC, self).__init__(x0, sigma0)
+
+        # Set initial state
+        self._x0 = np.asarray(x0, dtype=float)
+        self._running = False
+        self._ready_for_tell = False
+
+        # Iterations monitoring
+        self._mcmc_iteration = 0
+
+        # Current sample, log_pdf of the current sample
+        self._current = None
+        self._current_log_pdf = None
+
+        # Current log_y used to define the slice
+        self._current_log_y = None
+
+        # Current proposed sample
+        self._proposed = None
+
+        # Default initial interval width w used in the Stepout procedure
+        # to expand the interval
+        self._w = np.ones(len(self._x0))
+
+        # Default integer limiting the size of the interval to ``m*w```
+        self._m = 50
+
+        # Flag to initialise the expansion of the interval ``I=(L,R)``
+        self._first_expansion = False
+
+        # Flag indicating whether the interval expansion is concluded
+        self._interval_found = False
+
+        # Number of steps used for expanding the interval ``I=(L,R)``
+        self._j = None
+        self._k = None
+
+        # Edges of the interval ``I=(L,R)``
+        self._l = None
+        self._r = None
+
+        # Multi-dimensional points used to calculate the log_pdf of the edges ``l,r```
+        self._temp_l = None
+        self._temp_r = None
+
+        # Log_pdf of interval edge points ``l,r```
+        self._fx_l = None
+        self._fx_r = None
+
+        # Flags to indicate the interval edge to update
+        self._set_l = False
+        self._set_r  = False
+
+        # Index of parameter "xi" we are updating of the sample "x = (x1,...,xn)"
+        self._i = 0
+
+        # Stochastic variables as instance variables for testing
+        self._u = 0
+        self._v = 0
+        self._e = 0
+
+        # Flags used to calculate log_pdf of initial interval edges ``l,r```
+        self._init_left = False
+        self._init_right = False
+
+
+    def ask(self):
+        """ See :meth:`SingleChainMCMC.ask()`. """
+
+        # Check ask/tell pattern
+        if self._ready_for_tell:
+            raise RuntimeError('Ask() called when expecting call to tell().')
+
+        # Initialise on first call
+        if not self._running:
+            self._running = True
+
+        # Very first iteration
+        if self._current is None:
+
+            # Ask for the log pdf of x0
+            self._ready_for_tell = True
+            return np.array(self._x0, copy=True)
+
+
+        # If the flag is True, we initialise the expansion of interval ``I=(l,r)``
+        if self._first_expansion == True:
+
+            # Set initial values for l and r
+            self._u = np.random.uniform()
+            self._l = self._proposed[self._i] - self._w[self._i]*self._u
```

Comment:
Think that our style checker (flake8) needs spaces between operators. Here self._w[self._i] * self._u is what is needed, I think.

I also think we should use a more descriptive name for self._i. At the moment, it could be confused with an iteration counter, for example.